stats/thresholds_parser.go

```diff
@@ -0,0 +1,145 @@
+// gauge               -> "value"
+// rate                -> "rate"
+// trend               -> "avg" | "min" | "max" | "med" | percentile
+// percentile          -> "p(" float ")"
+// operator            -> ">" | ">=" | "<=" | "<" | "==" | "===" | "!="
+// float               -> digit+ ("." digit+)?
+// digit               -> "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9"
+// whitespace          -> " "
+// ```
+func parseThresholdExpression(input string) (*thresholdExpression, error) {
+	// Scanning makes no assumption on the underlying values, and only
+	// checks that the expression has the right format.
+	method, operator, value, err := scanThresholdExpression(input)
+	if err != nil {
+		return nil, fmt.Errorf("failed parsing threshold expression; reason: %w", err)
+	}
+
+	parsedMethod, err := parseThresholdAggregationMethod(method)
+	if err != nil {
+		return nil, fmt.Errorf("failed parsing threshold expression's left hand side; reason: %w", err)
+	}
+
+	parsedValue, err := strconv.ParseFloat(value, 64)
+	if err != nil {
+		return nil, fmt.Errorf("failed parsing threshold expresion's right hand side; reason: %w", err)
+	}
+
+	condition := &thresholdExpression{
+		AggregationMethod: parsedMethod,
+		Operator:          operator,
+		Value:             parsedValue,
+	}
+
+	return condition, nil
+}
+
+// scanThresholdExpression scans a threshold condition expression of the
+// form: `aggregation_method operator value`. An invalid or unknown operator
+// will produce an error. However, no assertions regarding
+// either the left-hand side aggregation method nor the right-hand
+// side value will be made: they will be returned as is, only trimmed from
+// their spaces.
+func scanThresholdExpression(input string) (string, string, string, error) {
+	operators := []string{
+		"<=",
+		"<",
+		">=",
+		">",
+		"===",
+		"==",
+		"!=",
+	}
+
+	for _, op := range operators {
+		substrings := strings.SplitN(input, op, 2)
+		if len(substrings) == 2 {
+			return strings.TrimSpace(substrings[0]), op, strings.TrimSpace(substrings[1]), nil
+		}
+	}
+
+	return "", "", "", fmt.Errorf("malformed threshold expression")
+}
+
+// parseThresholdMethod will parse a threshold condition expression's method.
+// It assumes the provided input argument is already trimmed and cleaned up.
+// If it encounters a percentile method, it will parse it and verify it
+// boils down to an expression of the form: `p(float64)`, but will return
+// it verbatim, as a string.
+func parseThresholdAggregationMethod(input string) (string, error) {
+	methods := []string{
+		"count",
+		"rate",
+		"value",
+		"avg",
+		"min",
+		"max",
+		"med",
+	}
+
+	// Is the input one of the methods keywords?
+	for _, m := range methods {
+		if m == input {
+			return input, nil
+		}
+	}
+
+	// Otherwise, attempt to parse a percentile expression
+	if strings.HasPrefix(input, "p(") && strings.HasSuffix(input, ")") {
+		trimmed := strings.TrimSuffix(strings.TrimPrefix(input, "p("), ")")
```

Comment:
Nit: I would've written this as `input[2:len(input)-1]` possibly replacing 2 and 1 with `len("p(")` 

Not particularly certain my shorter version is more readable as it's shorter, or your longer version is better with it being more specific/expressive  

I wonder if some helper function
```
func TrimSides(input, prefix, suffix string) (result string, didTrim bool) 
```
Will not be beneficial, I think I am doing this in other places as well.

But regardless, just mentioning it, it's not a requirement to be fixed.

Reply:
My personal preference is to go with a more functional style and try to avoid "static" indices in code as much as I can, as they involve the overhead of thinking about what they mean.

I've added a private `trimDelimited(prefix, input, suffix string) string` helper function to hide the logic and allow a much clearer `trimDelimited("p(", input, ")")` call.